### ISSUE TITLE
Set ACL with `put/file` (S3)

### ISSUE THREAD

**Issue by MicahElliott** [greghendershott/aws#52]
The AWS CLI has a `cp` command that makes it very convenient to set an ACL. A very common usage is: `aws s3 --acl=public-read cp foo.txt s3://mybucket.example.com/bar/foo.txt`

I see you've [provided](http://planet.racket-lang.org/package-source/gh/aws.plt/1/5/planet-docs/manual/S3__Storage_.html#%28def._%28%28planet._s3..rkt._%28gh._aws..plt._1._5%29%29._put/file%29%29) a `put-acl` function, but it's hard to figure out what a valid x-expression would be to use with it. The output from `get-acl` is pretty big, so do you have to replicate such an xexpr to use with `put-acl`?

It seems it would be very convenient for `put/file` to take a keyword argument like `put/file b+p p #:acl "public-read"`.

**Comment by greghendershott** [comment]
Your link is to doc for the old PLaneT package, so first I want to make sure you know that's out of date. Instead see [the new package](http://pkgs.racket-lang.org/#[aws]) and [its doc](http://pkg-build.racket-lang.org/doc/aws/index.html).

---

> I see you've [provided](http://planet.racket-lang.org/package-source/gh/aws.plt/1/5/planet-docs/manual/S3__Storage_.html#%28def._%28%28planet._s3..rkt._%28gh._aws..plt._1._5%29%29._put/file%29%29) a `put-acl` function, but it's hard to figure out what a valid x-expression would be to use with it. The output from `get-acl` is pretty big, so do you have to replicate such an xexpr to use with `put-acl`?

I agree it's not convenient.

My doc links to [the AWS doc explaining their XML for ACLs](http://docs.aws.amazon.com/AmazonS3/latest/dev/S3_ACLs_UsingACLs.html). My assumption was this would be the most comprehensive (and future-proof) way to support S3 ACLs -- just pass the XML back and forth as xexprs.

And in all honesty, it saved me from having to fully learn about S3 ACLs to the point where I could figure out a more Rackety abstraction in which to wrap them. I figured it would be worse if I attempted that but got it wrong, vs. leaving it up to users for whom ACLs are important.

> It seems it would be very convenient for `put/file` to take a keyword argument like `put/file b+p p #:acl "public-read"`.

I agree that would be more convenient.

---

My use of S3 is limited these days. Even when I used S3 more heavily, it was mostly buckets with the default access. So I don't have any hands-on with S3 ACLs. I also don't have an opinion about how to make the 80% case more convenient, while still allowing the 20% case. (I'm not even sure what the 80% case is, although I imagine it includes "public read".)

TL;DR: I think it would be wiser for something like this to be a pull request from someone for whom this is important?

**Comment by mflatt** [comment]
@MicahElliott In case you still care, with `put/bytes`, you can supply a canned ACL as a 'x-amz-acl header. Maybe `put/file` should accept extra headers, too.
Pull request #53 adds an extra-headers argument to `put-acl`, so that it can be used like `(put-acl <object> #f (hash 'x-amz-acl "public-read"))`
